// spotify-web-sdk/build/models/episode/episode-simplified.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var image_1 = __importDefault(require("../common/image"));
var EpisodeSimplified = (function () {
    function EpisodeSimplified(json) {
        this.audioPreviewUrl = json.audio_preview_url;
        this.description = json.description;
        this.durationMs = json.duration_ms;
        this.explicit = json.explicit;
        this.externalUrls = json.external_urls;
        this.href = json.href;
        this.id = json.id;
        this.images = json.images.map(function (imageJson) { return new image_1.default(imageJson); });
        this.isExternallyHosted = json.is_externally_hosted;
        this.isPlayable = json.is_playable;
        this.languages = json.languages;
        this.name = json.name;
        this.releaseDate = json.release_date;
        this.releaseDatePrecision = json.release_date_precision;
        if (json.resume_point) {
            this.resumePoint = new ResumePoint(json.resume_point);
        }
        this.type = json.type;
        this.uri = json.uri;
    }
    return EpisodeSimplified;
}());
var ResumePoint = (function () {
    function ResumePoint(json) {
        this.fullyPlayed = json.fully_played;
        this.resumePositionMs = json.resume_position_ms;
    }
    return ResumePoint;
}());
exports.default = EpisodeSimplified;
