// spotify-web-sdk/build/models/show/show-simplified.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var copyright_1 = __importDefault(require("../common/copyright"));
var image_1 = __importDefault(require("../common/image"));
var ShowSimplified = (function () {
    function ShowSimplified(json) {
        this.availableMarkets = json.available_markets;
        this.copyrights = json.copyrights.map(function (copyrightJson) { return new copyright_1.default(copyrightJson); });
        this.description = json.description;
        this.explicit = json.explicit;
        this.externalUrls = json.external_urls;
        this.href = json.href;
        this.id = json.id;
        this.images = json.images.map(function (imageJson) { return new image_1.default(imageJson); });
        this.isExternallyHosted = json.is_externally_hosted;
        this.languages = json.languages;
        this.mediaType = json.media_type;
        this.name = json.name;
        this.publisher = json.publisher;
        this.type = json.type;
        this.uri = json.uri;
    }
    return ShowSimplified;
}());
exports.default = ShowSimplified;
